src/main/java/hudson/plugins/build_timeout/BuildStepWithTimeout.java

```diff
@@ -0,0 +1,153 @@
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.BuildStepMonitor;
+import hudson.tasks.Builder;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.tokenmacro.MacroEvaluationException;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.StaplerRequest;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+public class BuildStepWithTimeout extends Builder implements BuildStep {
+    private /* final */ BuildTimeOutStrategy strategy;
+    private BuildStep buildStep;
+    private final List<BuildTimeOutOperation> operationList;
+
+    @DataBoundConstructor
+    public BuildStepWithTimeout(BuildStep buildStep, BuildTimeOutStrategy strategy, List<BuildTimeOutOperation> operationList) {
+        this.strategy = strategy;
+        this.buildStep = buildStep;
+        if (operationList != null && !operationList.isEmpty()) {
+            this.operationList = operationList;
+        }
+        else {
+            this.operationList = Collections.emptyList();
+        }
+
+    }
+
+    public List<BuildTimeOutOperation> getOperationList() {
+        return operationList;
+    }
+
+
+    public BuildStep getBuildStep() {
+        return buildStep;
+    }
+
+    public void setBuildStep(BuildStep buildStep) {
+        this.buildStep = buildStep;
+    }
+
+    private long getTimeout(Run run, TaskListener listener) throws IOException, InterruptedException {
+        try {
+            return strategy.getTimeOut((AbstractBuild<?, ?>) run, (BuildListener) listener);
+        } catch (MacroEvaluationException e) {
+            e.printStackTrace(listener.getLogger());
+            listener.error("Can't evaluate timeout - timeout would be disabled");
+            return Long.MAX_VALUE;
+        }
+    }
+
+
+    public boolean perform(final Build<?,?> build, final Launcher launcher, final BuildListener listener) throws InterruptedException, IOException {
+        final Timer timer = new Timer("Timer-" + build.getDisplayName().replace(" ", "_"), true);
+        final AtomicBoolean stopped = new AtomicBoolean(false);
+        try {
+            final long delay = getTimeout(build, listener);
+
+            timer.schedule(new TimerTask() {
+                @Override
+                public void run() {
+                    if (operationList.isEmpty()) {
+                        new AbortOperation().perform(build, listener, delay);
+                    }
+
+                    for (BuildTimeOutOperation op : operationList) {
+                        op.perform(build, listener, delay);
+                    }
+                    stopped.set(true);
+                    timer.cancel();
+                }
+            }, delay);
+
+            return buildStep.perform((AbstractBuild) build, launcher, listener);
+        } catch(InterruptedException e) {
+            if (!stopped.get()) {
+                throw e;
+            }
```

Comment:
What happens if you don't interrupt this exception?
(I'm not sure this catch block is actually necessary)